File: depth_first_search/1530_number_of_good_leaf_nodes_pairs/1530_number_of_good_leaf_nodes_pairs.py
import collections

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        if not root: return 0
        self.res = 0
        def dfs(node, depth):
            if not node.left and not node.right: return collections.Counter([depth])
            l, r = collections.Counter(), collections.Counter()
            if node.left:
                l = dfs(node.left, depth + 1)
            if node.right:
                r = dfs(node.right, depth + 1)
            for k in l:
                for j in r:
                    if j + k <= distance +2 * depth:
                        self.res += l[k] * r[j]
            return l + r
        dfs(root, 0)
        return self.res

File: depth_first_search/1530_number_of_good_leaf_nodes_pairs/1530_number_of_good_leaf_nodes_pairs.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func dfs(n *TreeNode, d int) (map[int]int, int) {
	if n == nil {
		return map[int]int{}, 0
	}
	if n.Left == nil && n.Right == nil {
		return map[int]int{0: 1}, 0
	}

	lp, lc := dfs(n.Left, d)
	rp, rc := dfs(n.Right, d)

	m := map[int]int{}
	c := lc + rc
	for kr, vr := range rp {
		m[kr+1] += vr
	}

	for kl, vl := range lp {
		m[kl+1] += vl
		for kr, vr := range rp {
			if kl+kr+2 <= d {
				c += vl * vr
			}
		}
	}

	//fmt.Println(n.Val, m, c)

	return m, c
}

func countPairs(root *TreeNode, distance int) int {
	_, lc := dfs(root, distance)
	return lc
}
